File: src/Runtime/Value/ValueObject.cs
/*
	PeakScript: https://github.com/Jenocn/PeakScript
	By Jenocn: https://jenocn.github.io
*/

namespace peak.interpreter
{

	public class ValueObject : TypeValue<ValueObject>
	{
		private Space _space = null;
		public Space space { get => _space; }
		public ValueObject(Space usingSpace = null, ValueObject parent = null)
		{
			Space parentSpace = parent ? parent.space.CopySpace() : null;
			_space = new Space(SpaceType.Object, parentSpace);
			if (usingSpace)
			{
				_space.AddSpaceOfUsing(usingSpace);
			}
		}
		public override string ToString()
		{
			string ret = "object { ";
			var variables = _space.variables;
			int index = 0;
			foreach (var item in variables)
			{
				ret += item.Value.name;
				ret += "=";
				ret += item.Value.value.ToString();
				if (index < variables.Count - 1)
				{
					ret += ", ";
				}
				index += 1;
			}
			ret += " }";
			return ret;
		}
		public override Value Clone()
		{
			return new ValueObject(_space.CopySpace());
		}
	}
} // namespace peak.interpreter

File: src/runtime/value/value_object.cpp
#include "value_object.h"
#include "runtime/space.h"
#include "runtime/variable.h"

using namespace peak;

ValueObject::ValueObject(std::shared_ptr<Space> indexSpace, std::shared_ptr<ValueObject> parent) {
	auto parentSpace = parent ? parent->GetSpace()->CopySpace() : nullptr;
	_space = std::make_shared<Space>(SpaceType::Object, parentSpace);
	if (indexSpace) {
		_space->AddSpaceOfUsing(indexSpace);
	}
}

std::shared_ptr<Value> ValueObject::Clone() const {
	auto ret = std::make_shared<ValueObject>();
	ret->_space = _space->CopySpace();
	return ret;
}

std::shared_ptr<Space> ValueObject::GetSpace() const {
	return _space;
}

std::string ValueObject::ToString() const {
	std::string ret = "object { ";
	const auto& variables = _space->GetVariables();
	auto index = 0u;
	for (auto& item : variables) {
		ret += item.second->GetName();
		ret += "=";
		ret += item.second->GetValue()->ToString();
		if (index < variables.size() - 1) {
			ret += ", ";
		}
		index += 1;
	}
	ret += " }";
	return ret;
}
